zqd/api/logwriter.go

```diff
@@ -0,0 +1,115 @@
+package api
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"io"
+	"mime/multipart"
+	"sync"
+	"sync/atomic"
+
+	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/zio/ndjsonio"
+)
+
+type LogWriter struct {
+	URIs       []iosrc.URI
+	BytesTotal int64
+
+	bytesRead int64
+	form      *multipart.Writer
+	pr        *io.PipeReader
+	pw        *io.PipeWriter
+	start     sync.Once
+	json      *ndjsonio.TypeConfig
+}
+
+func NewLogWriter(files ...string) (*LogWriter, error) {
+	pr, pw := io.Pipe()
+	form := multipart.NewWriter(pw)
+	lw := &LogWriter{form: form, pr: pr, pw: pw}
+	for _, f := range files {
+		u, err := iosrc.ParseURI(f)
+		if err != nil {
+			return nil, err
+		}
+		info, err := iosrc.Stat(context.Background(), u)
+		if err != nil {
+			return nil, err
+		}
+		lw.BytesTotal += info.Size()
+		lw.URIs = append(lw.URIs, u)
+	}
+	return lw, nil
+}
+
+func (l *LogWriter) SetJSONTyping(config *ndjsonio.TypeConfig) {
+	l.json = config
+}
+
+func (l *LogWriter) ContentType() string {
+	return l.form.FormDataContentType()
+}
+
+func (l *LogWriter) BytesRead() int64 {
+	return atomic.LoadInt64(&l.bytesRead)
+}
+
+func (l *LogWriter) Read(b []byte) (int, error) {
+	l.start.Do(func() {
+		go l.run()
+	})
+	return l.pr.Read(b)
+}
+
+func (l *LogWriter) run() {
+	if err := l.sendJSONTyping(); err != nil {
+		l.pw.CloseWithError(err)
+		return
+	}
+	for _, u := range l.URIs {
+		if err := l.writeFile(u); err != nil {
+			l.pw.CloseWithError(err)
+			return
+		}
+	}
+	l.pw.CloseWithError(l.form.Close())
+}
+
+func (l *LogWriter) writeFile(u iosrc.URI) error {
+	r, err := iosrc.NewReader(context.Background(), u)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	w, err := l.form.CreateFormFile("", u.String())
+	if err != nil {
+		return err
+	}
+	counter := &partCounter{reader: bufio.NewReader(r), nread: &l.bytesRead}
+	_, err = io.Copy(w, counter)
+	return err
+}
+
+func (l *LogWriter) sendJSONTyping() error {
```

Comment:
Nit: Again, I think we should drop the "typing" terminology entirely and call this sendJSONTypeConfig or sendJSONConfig.